internal/wasm/store.go

```diff
@@ -482,56 +501,69 @@ func (s *Store) Instantiate(module *Module, name string) (*PublicModule, error)
 	return &PublicModule{s: s, instance: instance}, nil
 }
 
-// ReleaseModuleInstance deallocates resources if a module with the given name exists.
-func (s *Store) ReleaseModuleInstance(moduleName string) error {
-	// TODO: none of this is goroutine safe
-	instance := s.moduleInstances[moduleName]
-	if instance == nil {
+// ReleaseModule deallocates resources if a module with the given name exists.
+func (s *Store) ReleaseModule(moduleName string) error {
+	m := s.module(moduleName)
+	if m == nil {
 		return nil // already released
 	}
 
-	instance.refCount--
-	if instance.refCount > 0 {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	if m.dependentCount > 0 {
 		// This case other modules are importing this module instance and still alive.
-		return fmt.Errorf("%d modules import this and need to be closed first", instance.refCount)
+		return fmt.Errorf("%d modules import this and need to be closed first", m.dependentCount)
 	}
 
 	// TODO: check outstanding calls and wait until they exit.
 
-	// Recursively release the imported instances.
-	for mod := range instance.moduleImports {
-		if err := s.ReleaseModuleInstance(mod.Name); err != nil {
-			return fmt.Errorf("unable to release imported module [%s]: %w", mod.Name, err)
-		}
+	for mod := range m.dependencies {
+		mod.decImportedCount()
 	}
 
-	if err := s.releaseFunctionInstances(instance.Functions...); err != nil {
+	if err := s.releaseFunctions(m.Functions...); err != nil {
 		return fmt.Errorf("unable to release function instance: %w", err)
 	}
 
-	if instance.MemoryInstance != nil {
-		s.releaseMemoryInstance(instance.MemoryInstance)
+	if m.Memory != nil {
+		s.releaseMemory(m.Memory)
 	}
 
-	if instance.TableInstance != nil {
-		s.releaseTableInstance(instance.TableInstance)
+	if m.Table != nil {
+		s.releaseTable(m.Table)
 	}
 
-	s.releaseGlobalInstances(instance.Globals...)
+	s.releaseGlobal(m.Globals...)
 
-	// Explicitly assign nil so that we ensure this moduleInstance no longer holds reference to instances.
-	instance.Exports = nil
-	instance.Globals = nil
-	instance.Functions = nil
-	instance.TableInstance = nil
-	instance.MemoryInstance = nil
-	instance.Types = nil
+	// Explicitly assign nil so that we ensure this module no longer holds reference to instances.
```

Comment:
mention why it is important especially as we clear the reference below.. is it because we want things that have this reference to fail (panic)?

Reply:
deleted as this is unnecessary